staging/src/k8s.io/client-go/util/csaupgrade/upgrade.go

```diff
@@ -0,0 +1,211 @@
+		func(entry metav1.ManagedFieldsEntry) bool {
+			return entry.Manager == ssaManagerName &&
+				entry.Operation == metav1.ManagedFieldsOperationApply &&
+				entry.Subresource == ""
+		})
+
+	if ssaManagerExists {
+		err = unionManagerIntoIndex(managedFields, ssaManagerIndex, csaManagerName)
+		if err != nil {
+			return nil, err
+		}
+
+	} else {
+		// SSA manager does not exist. Find the most recent matching CSA manager,
+		// convert it to an SSA manager.
+		//
+		// (find first index, since managed fields are sorted so that most recent is
+		//  first in the list)
+		csaManagerIndex, csaManagerExists := findFirstIndex(managedFields,
+			func(entry metav1.ManagedFieldsEntry) bool {
+				return entry.Manager == csaManagerName &&
+					entry.Operation == metav1.ManagedFieldsOperationUpdate &&
+					entry.Subresource == ""
+			})
+
+		if !csaManagerExists {
+			// There are no CSA managers that need to be converted. Nothing to do
+			// Return early
+			return obj, nil
+		}
+
+		// Convert CSA manager into SSA manager
+		managedFields[csaManagerIndex].Operation = metav1.ManagedFieldsOperationApply
+		managedFields[csaManagerIndex].Manager = ssaManagerName
+
+		err = unionManagerIntoIndex(managedFields, csaManagerIndex, csaManagerName)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Create version of managed fields which has no CSA managers with the given name
+	filteredManagers := filter(managedFields, func(entry metav1.ManagedFieldsEntry) bool {
+		return !(entry.Manager == csaManagerName &&
+			entry.Operation == metav1.ManagedFieldsOperationUpdate &&
+			entry.Subresource == "")
+	})
+
+	copied := obj.DeepCopyObject()
+	copiedAccessor, err := meta.Accessor(copied)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get meta accessor for copied object: %w", err)
+	}
+	copiedAccessor.SetManagedFields(filteredManagers)
+
+	// Wipe out CSA annotation if it exists
+	annotations := copiedAccessor.GetAnnotations()
+	delete(annotations, csaAnnotationName)
+	copiedAccessor.SetAnnotations(annotations)
+
+	return copied, nil
+}
+
+// Locates an Update manager entry named `csaManagerName` with the same APIVersion
+// as the manager at the targetIndex. Unions both manager's fields together
+// into the manager specified by `targetIndex`. No other managers are modified.
+func unionManagerIntoIndex(entries []metav1.ManagedFieldsEntry, targetIndex int, csaManagerName string) error {
+	ssaManager := entries[targetIndex]
+
+	// find Update manager of same APIVersion, union ssa fields with it.
+	// discard all other Update managers of the same name
+	csaManagerIndex, csaManagerExists := findFirstIndex(entries,
+		func(entry metav1.ManagedFieldsEntry) bool {
+			return entry.Manager == csaManagerName &&
+				entry.Operation == metav1.ManagedFieldsOperationUpdate &&
+				entry.Subresource == "" &&
+				entry.APIVersion == ssaManager.APIVersion
+		})
+
+	targetFieldSet, err := fieldsToSet(*ssaManager.FieldsV1)
```

Comment:
I know for now we only have one version (`FieldsV1`), can we make the function into ...
`ManagedEntryToSet(ssaManager)`? so that in the future if we have `FieldsV1` and `FieldsV2`, we don't have to change that specific code?

Reply:
done